### DISCUSSION TITLE
ACID & strict serializability versus best effort on a per request basis

### DISCUSSION THREAD

**Discussion by dumblob** [erthink/libmdbx#339] [Q&A] [answered]
### Motivation

Nowadays many setups do actually require a seamless combination of "ACID" key-value pairs and "best effort" key-value pairs (imagine analogy TCP packets and UDP packets treatment) in the **same distributed namespace/storage** (allowing e.g. transactions over a mix of "ACID" and "best effort" k-v pairs). The goal is to achieve the very minimal possible latency for the "best effort" k-v pairs and [strict serializability](https://jepsen.io/consistency ) and full durability for all other k-v pairs.

A naive DB interface for DB clients would then support a "tag" to distinguish between an "ACID" request on a key-value pair (or on a set of pairs) and a "best effort" request on a key-value pair (or on a set of pairs)?

For requests tagged with "best effort", the DB would e.g. ignore some ACID invariants or in the very worst case drop the whole request.

### Question: Is it possible to achieve similar behavior with the code from `master`? How?

**Comment by erthink** [comment, marked as answer]
Briefly:
1. MDBX always provide [serializability](https://en.wikipedia.org/wiki/Serializability). This can't altered or undone, but you could implement in-process batching for "best effort" requests, and "nested write transactions" feature could be useful for such cases.
2. MDBX affords wait-free for readers with read-committed isolation. This can't altered or undone, but you could implement in-process dirty reads (i.e. uncomited) for "best effort" requests.
3. Only "D" (i.e. the durability) in "ACID" could be relaxed by using MDBX API in a regular way. Please see [SYNC MODES](https://libmdbx.dqdkfa.ru/group__sync__modes.html) for more information.

**Comment by dumblob** [comment]
Thanks! That makes it clear where to draw the line.

> 3. Only "D" (i.e. the durability) in "ACID" could be relaxed by using MDBX API in a regular way. Please see [SYNC MODES](https://erthink.github.io/libmdbx/group__sync__modes.html) for more information.

I saw that but was unable to extract the information whether I can change sync mode "any time" (especially after the DB was opened or a transaction started etc.) - i.e. per-request and if so, whether there are any side effects (like poorer performance due to some caches issues or other inefficiencies).

Could you maybe comment on this as well?

**Comment by erthink** [comment]
> I saw that but was unable to extract the information whether I can change sync mode "any time" (especially after the DB was opened or a transaction started etc.) - i.e. per-request and if so, whether there are any side effects (like poorer performance due to some caches issues or other inefficiencies).

By [`mdbx_txn_begin_ex()`](https://libmdbx.dqdkfa.ru/group__c__transactions.html#ga2cfa3a3e97def47cafb4bc9b9e13da3d) you able to relax durability to `MDBX_NOMETASYNC` or `MDBX_SAFE_NOSYNC` for any non-nested write transaction.

**Comment by dumblob** [comment]
> By [`mdbx_txn_begin_ex()`](https://erthink.github.io/libmdbx/group__c__transactions.html#ga2cfa3a3e97def47cafb4bc9b9e13da3d) you able to relax durability to `MDBX_NOMETASYNC` or `MDBX_SAFE_NOSYNC` for any non-nested write transaction.

Does this affect other threads & processes which run their *read* transactions at the same time with different `MDBX...SYNC` options than the current writer?

It shouldn't but with vast majority of DBs always trying to trade safety for performance I can never be sure :wink:.

**Comment by erthink** [comment]
> Does this affect other threads & processes which run their _read_ transactions at the same time with different `MDBX...SYNC` options than the current writer?

No affects, but in case of system/power failure you may lose all non-sycn'ed (unsteady) changes/transactions.

Nonetheless, please read all about the [MDBX_SAFE_NOSYNC](https://libmdbx.dqdkfa.ru/group__c__opening.html#gga9138119a904355d245777c4119534061a586d75ff2fd68aed647a60ea7984ed85) with references, at least three times.
